catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/ws/JsonRpc.java

```diff
@@ -0,0 +1,176 @@
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import org.boon.json.JsonFactory;
+import org.boon.json.JsonParserFactory;
+import org.boon.json.JsonSerializerFactory;
+import org.boon.json.ObjectMapper;
+import org.eclipse.jetty.websocket.api.Session;
+
+// An implementation of http://www.jsonrpc.org/specification over websockets
+public class JsonRpc implements Socket {
+
+  private static String VERSION = "2.0";
+
+  private static Map<String, Object> METHOD_NOT_FOUND = error(-32601, "Method not found");
+  private static Map<String, Object> INVALID_PARAMS = error(-32602, "Invalid params");
+  private static Map<String, Object> INTERNAL_ERROR = error(-32603, "Internal error");
+
+  private ObjectMapper mapper =
+      JsonFactory.create(
+          new JsonParserFactory().usePropertyOnly(),
+          new JsonSerializerFactory()
+              .includeEmpty()
+              .includeNulls()
+              .includeDefaultValues()
+              .setJsonFormatForDates(false));
+
+  private Map<String, Function> methods;
+
+  public JsonRpc(Map<String, Function> methods) {
+    this.methods = methods;
+  }
+
+  private static Map<String, Object> parseError(Object message) {
+    return error(-32700, "Parse error", message);
+  }
+
+  private static Map<String, Object> error(int code, String message) {
+    return error(code, message, null);
+  }
+
+  private static Map<String, Object> error(int code, String message, Object data) {
+    Map<String, Object> response = new HashMap<>();
+    response.put("code", code);
+    response.put("message", message);
+    if (data != null) {
+      response.put("data", data);
+    }
+    return response;
+  }
+
+  private static Map<String, Object> success(Object id, Object result) {
+    Map<String, Object> response = new HashMap<>();
+    response.put("jsonrpc", VERSION);
+    response.put("id", id);
+    response.put("result", result);
+    return response;
+  }
+
+  private static Map<String, Object> fail(Object id, Object error) {
+    Map<String, Object> response = new HashMap<>();
+    response.put("jsonrpc", VERSION);
+    response.put("id", id);
+    response.put("error", error);
+    return response;
+  }
+
+  private Map<String, Object> invalid() {
+    return invalid(null);
+  }
+
+  private Map<String, Object> invalid(Object data) {
+    return error(-32600, "Invalid Request", data);
+  }
+
+  private Map<String, Object> invalidParams(Object data) {
+    return error(-32602, "Invalid params", data);
+  }
+
+  @Override
+  public void onOpen(Session session) {}
+
+  @Override
+  public void onClose(Session session, int statusCode, String reason) {}
+
+  @Override
+  public void onError(Session session, Throwable ex) {}
+
+  private Object exec(Object message) {
+
+    if (message instanceof List) {
+      return ((List) message).parallelStream().map(this::exec).collect(Collectors.toList());
```

Comment:
We should probably use something other than the global fork join pool for these operations

Reply:
custom threadpool for java parallel streams https://stackoverflow.com/a/22269778

Reply:
Just be aware that the parallelism is still determined by ForkJoinPool.getCommonPoolParallelism() from the common fork-join pool not by the new fork-join pool when you do this. (maybe comment it as such so anyone who copies the code also knows of this issue)

Reply:
What is the advantage of using a custom thread pool?